cmake/BuildLibSanitizers.cmake

```diff
@@ -23,36 +22,59 @@ macro(GEN_BUILD_TARGET name libsanitizer_path libfuzzer_path
 
   # Strip preinit object files in static libraries, otherwise a message
   # `.preinit_array section is not allowed in DSO` will prevent building DSO.
+  if(strip AND NOT "${strip}" STREQUAL "")
+    set(STRIP_COMMAND ${CMAKE_C_COMPILER_AR} d ${libsanitizer_name} ${strip})
+    set(STRIP_COMMENT "Strip sanitizer library ${name}")
+    set(STRIP_BYPRODUCTS BYPRODUCTS ${libsanitizer_name})
+  else()
+    set(STRIP_COMMAND ${CMAKE_COMMAND} -E echo "Skip stripping sanitizer library ${name} - no objects to strip")
+    set(STRIP_COMMENT "Skip stripping sanitizer library ${name} (no objects to strip)")
+    set(STRIP_BYPRODUCTS "")
+  endif()
+
   add_custom_target(strip_lib_${name}
-    COMMENT "Strip sanitizer library ${name}"
-    COMMAND ${CMAKE_C_COMPILER_AR} d ${libsanitizer_name} ${strip}
+    COMMENT ${STRIP_COMMENT}
+    COMMAND ${STRIP_COMMAND}
     WORKING_DIRECTORY ${CMAKE_CURRENT_BINARY_DIR}
-    BYPRODUCTS ${libsanitizer_name}
+    ${STRIP_BYPRODUCTS}
     DEPENDS copy_libs_${name}
   )
 
+  if (CMAKE_SYSTEM_NAME STREQUAL "Darwin")
+    set(LINK_COMMAND ${CMAKE_C_COMPILER} -Wl,-force_load,${libfuzzer_name}
+       -Wl,-force_load,${libsanitizer_name} -lstdc++ -lpthread
+       -dynamiclib -o ${sanitizer_dso_name})
+  else()
+    set(LINK_COMMAND ${CMAKE_C_COMPILER} -Wl,--whole-archive ${libfuzzer_name}
+       ${libsanitizer_name} -Wl,--no-whole-archive -lstdc++ -lpthread -ldl
+       -shared -o ${sanitizer_dso_name})
+  endif()
+
   add_custom_target(build_dso_${name} ALL
     COMMENT "Build sanitizer library ${name}"
-    COMMAND ${CMAKE_C_COMPILER} -Wl,--whole-archive ${libfuzzer_name}
-       ${libsanitizer_name} -Wl,--no-whole-archive -lstdc++ -lpthread -ldl
-       -shared -o ${sanitizer_dso_name}
+    COMMAND ${LINK_COMMAND}
     WORKING_DIRECTORY ${CMAKE_CURRENT_BINARY_DIR}
     BYPRODUCTS ${sanitizer_dso_name}
     DEPENDS strip_lib_${name}
   )
 endmacro()
 
-list(APPEND LIBCLANG_ASAN_STRIP
-  asan_preinit.cc.o
-  asan_preinit.cpp.o
-)
-list(APPEND LIBCLANG_UBSAN_STRIP
-  ubsan_init_standalone_preinit.cc.o
-  ubsan_init_standalone_preinit.cpp.o
-)
+if (CMAKE_SYSTEM_NAME STREQUAL "Darwin")
+  list(APPEND LIBCLANG_ASAN_STRIP "")
+  list(APPEND LIBCLANG_UBSAN_STRIP "")
+else()
+  list(APPEND LIBCLANG_ASAN_STRIP
+    asan_preinit.cc.o
+    asan_preinit.cpp.o
+  )
+  list(APPEND LIBCLANG_UBSAN_STRIP
+    ubsan_init_standalone_preinit.cc.o
+    ubsan_init_standalone_preinit.cpp.o
+  )
+endif()
```

Comment:
This would be better:

```
set(LIBCLANG_ASAN_STRIP "")
set(LIBCLANG_UBSAN_STRIP "")
if (CMAKE_SYSTEM_NAME STREQUAL "Linux")
  list(APPEND LIBCLANG_ASAN_STRIP
    asan_preinit.cc.o
    asan_preinit.cpp.o
  )
  list(APPEND LIBCLANG_UBSAN_STRIP
    ubsan_init_standalone_preinit.cc.o
    ubsan_init_standalone_preinit.cpp.o
  )
endif()
```